docs/testing-with-codeception.md

```diff
@@ -0,0 +1,746 @@
+
+#### Acceptance tests
+
+In short: *Testing a scenario from a user perspective, **in the browser***, ie: opening signup page, typing credentials, clicking sign in, and checking browser output.
+
+This type uses a browser, where a web driver drives the browser clicking and typing to simulate user actions.
+
+These can be written in _CEPT_ format, eg:
+
+```php
+// SignupSubmissionTest.php
+
+// Add a page that contains the shortcode that will render the signup form.
+$I->havePageInDatabase( [
+    'post_name' => 'signup',
+    'post_content'=> 'Sign-up for our awesome thing! [signup]',
+] );
+
+// Go to the page.
+$I->amOnPage( '/signup' );
+
+// Submit the form as a user would submit it.
+$I->submitForm( '#signup-form', [
+  'name' => 'John Doe',
+  'email' => 'john.doe@altis.com',
+] );
+
+// Make sure I see a confirmation message.
+$I->waitForElement( '#signup-confirmation' );
+```
+
+or the more nuanced _CEST_ format, largerly recommended due to its DRY capabilities, eg:
+
+```php
+// SignupSubmissionCest.php
+
+class SignupSubmissionCest {
+
+    public function _before( FunctionalTester $I ) {
+        // Add a page that contains the shortcode that will render the signup form.
+        $I->havePageInDatabase( [
+            'post_name' => 'signup',
+            'post_content'=> 'Sign-up for our awesome thing! [signup]',
+        ] );
+        $I->amOnPage( '/signup' );
+    }
+
+    public function test_good_signup( FunctionalTester $I ) {
+        // Submit the form as a user would submit it.
+        $I->submitForm( '#signup-form', [
+            'name' => 'John Doe',
+            'email' => 'john.doe@altis.com',
+        ] );
+
+        // Make sure I see a confirmation message.
+        $I->waitForElement( '#signup-confirmation' );
+    }
+
+    public function test_bad_email_signup( FunctionalTester $I ) {
+        // Submit the form as a user would submit it.
+        $I->submitForm( '#signup-form', [
+            'name' => 'John Doe',
+            'email' => 'not-really-an-email',
+        ] );
+
+        // Make sure I see an error message.
+        $I->waitForElement( '#signup-error' );
+    }
+}
+```
+
+#### Functional tests
+
+In short: *Testing a scenario from a developer perspective*, eg: sending AJAX/API requests and checking responses and/or database changes.
+
+This type doesn't necessarily use a browser, as it can use a PHP library that acts like a browser, with no Javascript support. And it looks pretty much like Acceptance tests, but with different testing mindset/domain.
+
+Functional and acceptance tests can theoretically/technically co-exist, eg: executing browser actions and checking expected database changes rather than just browser outout. But it adds the overhead of executing tests via a browser which can be slower than the quicker PHP browser.
```

Comment:
```suggestion
Functional and acceptance tests can co-exist, eg: executing browser actions and checking expected database changes rather than just browser output. But it adds the overhead of executing tests via a browser which can be slower than the quicker PHP browser.
```

Hmm, I'm not sure this paragraph really explains the difference, I find it a bit confusing. Maybe we could simplify it down to listing the key differences:

- No JavaScript support
- Much faster to run
- Ideal for testing static output such as API responses

Reply:
Yeah, the last sentence doesn't seem to fit, since acceptance tests can still be carried out by WPBrowser, not necessarily WPWebDriver. Will remove it entirely. The meaning is conveyed somewhere else already, no need to reiterate.